Only get processes info if pid is running [fix #1303]

marpaia opened the pull request.

mofarrell commented:
This is also an issue in linux builds.

marpaia commented:
@mofarrell good catch, i'll add it for linux as well.

theopolis commented:
(1) What is the motivation for the additional header/implementation that is shared among both Linux/Darwin platforms? The implementation seems very platform specific such that it'll be more useful to just include alongside either's `processes.cpp`, which already contains lots of useful process-related utility methods. 
(2) Are we sure the processes table is the only one affected?

mofarrell commented:
Just wanted to revoice my two concerns from https://github.com/marpaia/osquery/commit/493a66bd709bf45544177fd6b34b3a421672e56a#comments

marpaia commented:
@theopolis agreed with 1, I changed them back to occur in the individual table implementations. With regards to 2, I updated the other process related table by moving out the pid acquisition into a utility and using it from the table generators for the other tables.

@mofarrell gotcha, fair point, I added a constraint before we iterate the pid list that ensures that the size of the constraints is less than the total active pids. This is easy on OS X, because we have the call to `proc_listpids` which returns the count of items without returning the actual pids. On linux, we're just parsing `/proc`, so there's a bit of a caveat. I thought about adding another function which returns the count of pids, but since the implementation would be exactly the same complexity, I think it makes sense to count the number of pids in `/proc` by normally iterating `/proc`, then compare the `size()` of the `set` with the number of elements in the constraint list. if there are in fact less constraints, we clear the `set`, re-using the same data structure and re-fill it with the smaller, verified subset of pids.

marpaia commented:
@osqueryer test this please

theopolis commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +56,26 @@ inline std::string readProcLink(const std::string& attr, const std::string& pid)
 }
 
+Status getPidsForContext(QueryContext& context, std::set<std::string>& pids) {
+  auto s = osquery::procProcesses(pids);
```
I'm interested in the latency incurred by visiting all leafs in the `/proc` directory.

theopolis commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +56,26 @@ inline std::string readProcLink(const std::string& attr, const std::string& pid)
+    pids.clear();
+    for (const auto& pid :
+         context.constraints["pid"].getAll<std::string>(EQUALS)) {
+      if (isDirectory("/proc/" + pid).ok()) {
```
Don't you already know if the pid was in the `pids` set? What if you construct a new set if the constraint exists and move matching pids over? That will be considerably less work than `stat`ing the directory.

theopolis commented on osquery/tables/system/darwin/processes.cpp:
```diff
@@ -33,11 +33,40 @@ namespace tables {
-    pidlist = context.constraints.at("pid").getAll<int>(EQUALS);
+      context.constraints.at("pid").exists(EQUALS) &&
+      context.constraints.at("pid").getAll<int>(EQUALS).size() <
+          (bufsize / sizeof(pid_t))) {
```
Maybe a `auto& constraint = context.constraints.at("pid")` for a more readable `if` statement? Might not make a difference.
